File: 3907-CountPrimeGapBalancedSubarrays/3907-CountPrimeGapBalancedSubarrays.java
// Last updated: 7/21/2026, 11:50:34 AM
class Solution {
    private boolean[] prime;

    public int primeSubarray(int[] nums, int k) {
        int n = nums.length, max = 0;

        //  maximum value in nums for sieve limit
        for (int num : nums) max = Math.max(max, num);

        TreeMap<Integer, Integer> map = new TreeMap<>(); //sorted order track
        int primes = 0;

        // l1 -left ptr with atmmost one prime (invalid subarray counts so atlast we can substract that from total count)
        // l2 -> checking preime gap <=k
        int l1 = 0, l2 = 0;

        sieve(max);

        // cnt 1- count subarray atlmost one prime
        // cntK - count subarray with prime gap <= k
        int cntK = 0, cnt1 = 0;
        for (int r = 0; r < n; r++) {
            if (prime[nums[r]]) { //check if prime then put in window
                primes++;
                map.put(nums[r], map.getOrDefault(nums[r], 0) + 1);
            }

            while (primes > 1) {
                if (prime[nums[l1]]) primes--;
                l1++;
            }

            // Maintain window [l2..r] with (maxPrime - minPrime) <= k
            while (!map.isEmpty() && map.lastKey() - map.firstKey() > k) {
                if (prime[nums[l2]]) {
                    map.put(nums[l2], map.get(nums[l2]) - 1);
                    if (map.get(nums[l2]) == 0) map.remove(nums[l2]);
                }
                l2++;
            }

            
            cnt1 += (r - l1 + 1); //valid subarrays with atleast 2 prime 
            cntK += (r - l2 + 1); //maximum possible with gap of <=k 
        }
        return cntK - cnt1;
    }

    private void sieve(int n) {
        prime = new boolean[n + 1];
        Arrays.fill(prime, true);
        prime[1] = false;

        for (int p = 2; p * p <= n; p++) {
            if (prime[p]) {
                for (int i = p * p; i <= n; i += p)
                    prime[i] = false;
            }
        }
    }
}

File: 3907-CountPrimeGapBalancedSubarrays/3907-CountPrimeGapBalancedSubarrays.cpp
// Last updated: 7/21/2026, 11:50:28 AM
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<bool> prime;

    int primeSubarray(vector<int>& nums, int k) {
        int n = nums.size(), mx = 0;

        // maximum value in nums for sieve limit
        for (int num : nums) mx = max(mx, num);

        map<int, int> mp; // same as TreeMap (sorted)
        int primes = 0;

        int l1 = 0, l2 = 0;

        sieve(mx);

        int cntK = 0, cnt1 = 0;

        for (int r = 0; r < n; r++) {
            if (prime[nums[r]]) {
                primes++;
                mp[nums[r]]++;
            }

            while (primes > 1) {
                if (prime[nums[l1]]) primes--;
                l1++;
            }

            // Maintain window [l2..r] with (maxPrime - minPrime) <= k
            while (!mp.empty() && (mp.rbegin()->first - mp.begin()->first > k)) {
                if (prime[nums[l2]]) {
                    mp[nums[l2]]--;
                    if (mp[nums[l2]] == 0) mp.erase(nums[l2]);
                }
                l2++;
            }

            cnt1 += (r - l1 + 1);
            cntK += (r - l2 + 1);
        }

        return cntK - cnt1;
    }

    void sieve(int n) {
        prime.assign(n + 1, true);
        if (n >= 0) prime[0] = false;
        if (n >= 1) prime[1] = false;

        for (int p = 2; p * p <= n; p++) {
            if (prime[p]) {
                for (int i = p * p; i <= n; i += p) {
                    prime[i] = false;
                }
            }
        }
    }
};